File: Photos/app/src/main/java/com/hcmus/clc18se/photos/utils/DrawableImageView.java
package com.hcmus.clc18se.photos.utils;


import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;

import androidx.appcompat.widget.AppCompatImageView;

// link https://pristalovpavel.wordpress.com/2015/05/16/drawing-on-image-in-android/
public class DrawableImageView extends AppCompatImageView implements View.OnTouchListener {

    float downx = 0;
    float downy = 0;
    float upx = 0;
    float upy = 0;
    Canvas canvas;
    Paint paint;
    Matrix matrix;

    public DrawableImageView(Context context) {
        super(context);
        setOnTouchListener(this);
    }

    public DrawableImageView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setOnTouchListener(this);
    }

    public DrawableImageView(Context context, AttributeSet attrs,
                             int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setOnTouchListener(this);
    }

    public void setNewImage(Bitmap alteredBitmap, Bitmap bmp, ColorFilter colorFilter) {
        canvas = new Canvas(alteredBitmap);
        paint = new Paint();
        paint.setAntiAlias(true);
        paint.setDither(true);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_OVER));
        paint.setColor(Color.GREEN);
        paint.setStrokeWidth(5);
        paint.setColorFilter(colorFilter);
        matrix = new Matrix();
        canvas.drawBitmap(bmp, matrix, paint);

        setImageBitmap(alteredBitmap);
    }

    public void setNewColor(int color) {
        paint.setColor(color);
    }

    public void setWeight(int weight) {
        paint.setStrokeWidth(weight);
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        int action = event.getAction();

        switch (action) {
            case MotionEvent.ACTION_DOWN:
                downx = getPointerCoords(event)[0];//event.getX();
                downy = getPointerCoords(event)[1];//event.getY();
                break;
            case MotionEvent.ACTION_MOVE:
                upx = getPointerCoords(event)[0];//event.getX();
                upy = getPointerCoords(event)[1];//event.getY();
                canvas.drawLine(downx, downy, upx, upy, paint);
                invalidate();
                downx = upx;
                downy = upy;
                break;
            case MotionEvent.ACTION_UP:
                upx = getPointerCoords(event)[0];//event.getX();
                upy = getPointerCoords(event)[1];//event.getY();
                canvas.drawLine(downx, downy, upx, upy, paint);
                invalidate();
                break;
            case MotionEvent.ACTION_CANCEL:
                break;
            default:
                break;
        }
        return true;
    }

    final float[] getPointerCoords(MotionEvent e) {
        final int index = e.getActionIndex();
        final float[] coords = new float[]{e.getX(index), e.getY(index)};
        Matrix matrix = new Matrix();
        getImageMatrix().invert(matrix);
        matrix.postTranslate(getScrollX(), getScrollY());
        matrix.mapPoints(coords);
        return coords;
    }
}

File: Photos/app/src/main/java/com/hcmus/clc18se/photos/utils/DrawableImageView.kt
package com.hcmus.clc18se.photos.utils

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import androidx.appcompat.widget.AppCompatImageView

class DrawableImageView : AppCompatImageView, OnTouchListener {
    var downx = 0f
    var downy = 0f
    var upx = 0f
    var upy = 0f
    var canvas: Canvas? = null

    var paint: Paint? = null

    constructor(context: Context) : super(context) {
        setOnTouchListener(this)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        setOnTouchListener(this)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int
    ) : super(context, attrs, defStyleAttr) {
        setOnTouchListener(this)
    }

    fun setNewImage(alteredBitmap: Bitmap, bmp: Bitmap, colorFilter: ColorFilter?) {
        canvas = Canvas(alteredBitmap)

        paint = Paint().apply {
            isAntiAlias = true
            isDither = true
            style = Paint.Style.STROKE
            strokeJoin = Paint.Join.ROUND
            strokeCap = Paint.Cap.ROUND
            xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OVER)
            color = Color.GREEN
            strokeWidth = 5f
            this.colorFilter = colorFilter
        }

        imageMatrix = Matrix()
        canvas?.drawBitmap(bmp, matrix!!, paint)
        setImageBitmap(alteredBitmap)
    }

    fun setNewColor(color: Int) {
        paint?.color = color
    }

    fun setWeight(weight: Int) {
        paint?.strokeWidth = weight.toFloat()
    }

    override fun onTouch(v: View, event: MotionEvent): Boolean {

        when (event.action) {

            MotionEvent.ACTION_DOWN -> {
                downx = getPointerCoords(event)[0] //event.getX();
                downy = getPointerCoords(event)[1] //event.getY();
            }

            MotionEvent.ACTION_MOVE -> {
                upx = getPointerCoords(event)[0] //event.getX();
                upy = getPointerCoords(event)[1] //event.getY();
                canvas!!.drawLine(downx, downy, upx, upy, paint!!)
                invalidate()
                downx = upx
                downy = upy
            }

            MotionEvent.ACTION_UP -> {
                upx = getPointerCoords(event)[0] //event.getX();
                upy = getPointerCoords(event)[1] //event.getY();
                canvas!!.drawLine(downx, downy, upx, upy, paint!!)
                invalidate()
            }

            MotionEvent.ACTION_CANCEL -> { }

        }
        return true
    }

    private fun getPointerCoords(e: MotionEvent): FloatArray {
        val index = e.actionIndex
        val coords = floatArrayOf(e.getX(index), e.getY(index))
        val matrix = Matrix()

        imageMatrix.invert(matrix)
        matrix.postTranslate(scrollX.toFloat(), scrollY.toFloat())
        matrix.mapPoints(coords)
        return coords
    }
}